experimental/gradually-typed/src/Torch/GraduallyTyped/Tensor/MathOperations/Reduction.hs

```diff
@@ -48,65 +49,267 @@ type MeanErrorMessage (by :: By Symbol Nat) (dims :: [Dim (Name Symbol) (Size Na
       (names, indexes) = flip execState (Set.empty, Set.empty) $ do
         for_ bys' $ \by -> do
           case by of
             ByName name -> modify . first $ Set.insert name
             ByIndex index -> modify . second $ Set.insert index
-   in unsafePerformIO $ do
+   in unsafeThrowableIO $ do
         case (names, indexes) of
-          (names, indexes)
-            | Set.null names && Set.null indexes ->
+          (names', indexes')
+            | Set.null names' && Set.null indexes' ->
               do
                 t :: ForeignPtr ATen.Tensor <- cast tensor pure
                 uncast t pure
-            | Set.null names ->
-              cast1 (meanIndexes indexes) tensor
-            | Set.null indexes ->
-              cast1 (meanNames names) tensor
+            | Set.null names' ->
+              cast1 (meanIndexes indexes') tensor
+            | Set.null indexes' ->
+              cast1 (meanNames names') tensor
             | otherwise ->
               do
-                t' :: ForeignPtr ATen.Tensor <- cast1 (meanIndexes indexes) tensor
-                cast1 (meanNames names) t'
+                t' :: ForeignPtr ATen.Tensor <- cast1 (meanIndexes indexes') tensor
+                cast1 (meanNames names') t'
   where
     meanNames :: Set.Set String -> ForeignPtr ATen.Tensor -> IO (ForeignPtr ATen.Tensor)
-    meanNames names tensor =
+    meanNames names tensor' =
       cast3
         ATen.mean_tNb
-        tensor
+        tensor'
         (Set.toList names)
         True -- keepDim
     meanIndexes :: Set.Set Integer -> ForeignPtr ATen.Tensor -> IO (ForeignPtr ATen.Tensor)
-    meanIndexes indexes tensor =
+    meanIndexes indexes tensor' =
       cast3
         ATen.mean_tlb
-        tensor
+        tensor'
         (Set.toList indexes)
         True -- keepDim
 
-meanDim ::
+-- | Reduce the mean value over each row of the input tensor in the dimensions selected by 'selectDims'.
+-- For a version that accepts singleton parameters see 'sMeanDim'.
+--
+-- >>> g <- sMkGenerator (SDevice SCPU) 0
+-- >>> type Shape' = 'Shape '[ 'Dim ('Name "batch") ('Size 8), 'Dim ('Name "feature") ('Size 4) ]
+-- >>> (t, _) = randn @('Gradient 'WithoutGradient) @('Layout 'Dense) @('Device 'CPU) @('DataType 'Float) @Shape' g
+-- >>> :type meanDims @('SelectDims '[ 'ByName "feature" ]) t
+-- meanDims @('SelectDims '[ 'ByName "feature" ]) t
+--   :: Tensor
+--        ('Gradient 'WithoutGradient)
+--        ('Layout 'Dense)
+--        ('Device 'CPU)
+--        ('DataType 'Float)
+--        ('Shape
+--           '[ 'Dim ('Name "batch") ('Size 8),
+--              'Dim ('Name "feature") ('Size 1)])
+meanDims ::
   forall selectDims gradient layout device dataType shape.
   SingI selectDims =>
   Tensor gradient layout device dataType shape ->
   Tensor gradient layout device dataType (MeanF selectDims shape)
-meanDim = sMeanDim (sing @selectDims)
+meanDims = unsafePerformIO . sMeanDims (sing @selectDims)
```

Comment:
because of all this libtorch nonsense about bool and named dimensions and mean not being supported for some obscure tensor options, I'd leave it at `m (Tensor ...)`. this may still throw.